closed/src/java.base/share/classes/openj9/internal/security/RestrictedSecurity.java

```diff
@@ -0,0 +1,1044 @@
+
+    static {
+        String[] props = AccessController.doPrivileged(
+                new PrivilegedAction<>() {
+                    @Override
+                    public String[] run() {
+                        return new String[] { System.getProperty("semeru.fips"),
+                                System.getProperty("semeru.restrictedsecurity"),
+                                System.getProperty("os.name"),
+                                System.getProperty("os.arch") };
+                    }
+                });
+        userEnabledFIPS = Boolean.parseBoolean(props[0]);
+        // If semeru.fips is true, then ignore semeru.restrictedsecurity, use userSecurityNum 1.
+        userSecuritySetting = userEnabledFIPS ? "1" : props[1];
+        userEnabledSecurity = !isNullOrBlank(userSecuritySetting);
+        isSecuritySupported = "Linux".equalsIgnoreCase(props[2])
+                && supportPlatforms.contains(props[3]);
+        shouldEnableSecurity = (userEnabledFIPS || userEnabledSecurity) && isSecuritySupported;
+    }
+
+    private RestrictedSecurity() {
+        super();
+    }
+
+    /**
+     * Restricted security mode will be enabled when the semeru.restrictedsecurity
+     * system property is set.
+     *
+     * @return true if restricted security is enabled
+     */
+    public static boolean isEnabled() {
+        return securityEnabled;
+    }
+
+    /**
+     * Restricted security mode secure random provider
+     *
+     * @return the secure random provider
+     */
+    public static String getRandomProvider() {
+        if (!securityEnabled) {
+            printStackTraceAndExit(
+                    "Restricted security mode secure random provider can only be used when restricted security mode is enabled.");
+        }
+        return restricts.jdkSecureRandomProvider;
+    }
+
+    /**
+     * Restricted security mode secure random algorithm
+     *
+     * @return the secure random algorithm
+     */
+    public static String getRandomAlgorithm() {
+        if (!securityEnabled) {
+            printStackTraceAndExit(
+                    "Restricted security mode secure random algorithm can only be used when restricted security mode is enabled.");
+        }
+        return restricts.jdkSecureRandomAlgorithm;
+    }
+
+    /**
+     * FIPS mode will be enabled when the semeru.fips system property is true,
+     * or semeru.restrictedsecurity system property is set by using FIPS policy.
+     *
+     * @return true if FIPS is enabled
+     */
+    public static boolean isFIPSEnabled() {
+        return securityEnabled && (userSecurityNum == 1);
+    }
+
+    /**
+     * Check if the service is allowed in restricted security mode.
+     *
+     * @param service the service to check
+     * @return true if the service is allowed
+     */
+    public static boolean isServiceAllowed(Service service) {
+        if (securityEnabled) {
+            return restricts.isRestrictedServiceAllowed(service);
```

Comment:
I suggest the service class should be checked before delegating to `RestrictedSecurityProperties`.

Reply:
@keithc-ca 

In the Provider.java, before the “RestrictedSecurity.isServiceAllowed(s)” is called, there are the check for the service class, codes as below. So I did not add the check in the “RestrictedSecurity” class “isServiceAllowed()” method. 

```
        if (s == null) {
            throw new NullPointerException();
        }
        if (s.getProvider() != this) {
            throw new IllegalArgumentException
                    ("service.getProvider() must match this Provider object");
        }
        if (!RestrictedSecurity.isServiceAllowed(s)) {
            // We're in restricted security mode which does not allow this service,
            // return without registering.
            return;
        }
```

Reply:
@keithc-ca 

For the line 182, the method "isProviderAllowed(Class<?> providerClazz)". Before this method is called from the ServiceLoader.java, it had the checks for the providerClazz. Checks from ServiceLoader.java are below: 

```
        if (clazz == null) {
            fail(service, "Provider " + cn + " not found");
        }

        int mods = clazz.getModifiers();
        if (!Modifier.isPublic(mods)) {
            fail(service, clazz + " is not public");
        }

        if (!RestrictedSecurity.isProviderAllowed(clazz)) {
            // We're in restricted security mode which does not allow this provider,
            // skip it.
            return null;
        }
```

and 

```
                    Class<?> clazz = nextProviderClass();
                    if (clazz == null)
                        return false;

                    if (clazz.getModule().isNamed()) {
                        // ignore class if in named module
                        continue;
                    }

                    if (service.isAssignableFrom(clazz)) {
                        if (!RestrictedSecurity.isProviderAllowed(clazz)) {
                            // We're in restricted security mode which does not allow this provider,
                            // skip it.
                            continue;
                        }
```